packages/flutter_tools/lib/src/android/android_studio.dart

```diff
@@ -0,0 +1,257 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:pub_semver/pub_semver.dart';
+
+import '../base/common.dart';
+import '../base/context.dart';
+import '../base/file_system.dart';
+import '../base/os.dart';
+import '../base/platform.dart';
+import '../base/process_manager.dart';
+import '../globals.dart';
+import '../ios/plist_utils.dart';
+
+AndroidStudio get androidStudio =>
+    context.putIfAbsent(AndroidStudio, AndroidStudio.latestValid);
+
+// Android Studio layout:
+
+// Linux/Windows:
+// $HOME/.AndroidStudioX.Y/system/.home
+
+// macOS:
+// /Applications/Android Studio.app/Contents/
+// $HOME/Applications/Android Studio.app/Contents/
+
+// $STUDIO_HOME/gradle/gradle-X.Y.Z/bin/gradle
+
+final Version minGradleVersion = new Version(2, 14, 1);
+
+/// Locate Gradle.
+String get gradleExecutable {
+  // See if the user has explicitly configured gradle-dir.
+  String gradleDir = config.getValue('gradle-dir');
+  if (gradleDir != null) {
+    if (fs.isFileSync(gradleDir))
+      return gradleDir;
+    return fs.path.join(gradleDir, 'bin', 'gradle');
+  }
+  return androidStudio?.gradleExecutable ?? os.which('gradle')?.path;
+}
+
+class AndroidStudio implements Comparable<AndroidStudio> {
+  AndroidStudio(this.directory, {this.version = '0.0', this.configured}) {
+    _init();
+  }
+
+  final String directory;
+  final String version;
+  final String configured;
+
+  String _gradlePath;
+  bool _isValid = false;
+  List<String> _validationMessages = <String>[];
+
+  factory AndroidStudio.fromMacOSBundle(String bundlePath) {
+    String studioPath = fs.path.join(bundlePath, 'Contents');
+    String plistFile = fs.path.join(studioPath, 'Info.plist');
+    String version =
+        getValueFromFile(plistFile, kCFBundleShortVersionStringKey);
+    return new AndroidStudio(studioPath, version: version);
+  }
+
+  factory AndroidStudio.fromHomeDot(Directory homeDotDir) {
+    String version = homeDotDir.basename.substring('.AndroidStudio'.length);
+    String installPath;
+    try {
+      installPath = fs
+          .file(fs.path.join(homeDotDir.path, 'system', '.home'))
+          .readAsStringSync();
+    } catch (e) {
+      // ignored
```

Comment:
what error specifically are we ignoring here? We shouldn't blank-ignore all errors, only the harmless ones we know we're expecting.